### PR TITLE
Added Resume capability with the associated tests

### PR DESCRIPTION
**Pull request by harmjeff**
# Ralph Resume Feature - Complete Implementation Guide

## Overview

The `ralph resume` command allows users to restart a stopped Ralph loop with full context about previous work completed. This feature includes automatic state preservation on interruption, making resume truly functional.

## Table of Contents

- [Key Features](#key-features)
- [State Preservation System](#state-preservation-system)
- [Implementation Details](#implementation-details)
- [Usage Guide](#usage-guide)
- [Technical Architecture](#technical-architecture)
- [Files Changed](#files-changed)

---

### 1. Automatic State Preservation

Ralph now intelligently preserves loop state when interrupted:

| Event | State Preserved? | Can Resume? | User Experience |
|-------|-----------------|-------------|-----------------|
| **Ctrl+C (SIGINT)** | ✅ Yes | ✅ Yes | "Saving state for resume... Run 'ralph resume' to continue" |
| **Max iterations** | ✅ Yes | ✅ Yes | "Saving state for resume... Run 'ralph resume' to continue" |
| **`ralph cancel`** | ✅ Yes | ✅ Yes | "Run 'ralph resume' to continue where you left off" |
| **Successful completion** | ❌ No (deleted) | ❌ No | State cleaned up automatically |

### 2. Enhanced Resume Context

When resuming, the agent receives detailed context:
- Current iteration number
- Instructions to review files, git history, and test results
- Previous feedback (quality score, next steps, improvements, blockers, ideas)
- Original task prompt

### 3. Flexible Override Options

Resume with existing settings or override any parameter:
```bash
ralph resume                        # Use existing settings
ralph resume -m 30                  # Increase max iterations
ralph resume -p "FINISHED"          # Change completion promise
ralph resume -n 5 -m 25 -p "DONE"   # Override everything
```

---

#### Before (Broken Behavior)
```bash
# Start a loop
ralph loop "Build API" -m 10

# Hit Ctrl+C at iteration 5
# ❌ State file deleted
# ❌ No way to resume

ralph resume
# Error: "No stopped Ralph loop found"
```

#### After (Fixed Behavior)
```bash
# Start a loop
ralph loop "Build API" -m 10

# Hit Ctrl+C at iteration 5
# ✅ State file preserved with active: false
# ✅ Iteration and feedback saved

ralph resume
# ✅ Success! Resumes from iteration 5
# ✅ Shows previous feedback and context
```

### State File Structure

When preserved, the state file (`.kiro/ralph-loop.local.json`) contains:

```json
{
  "active": false,                    // Marked inactive, ready for resume
  "iteration": 12,                    // Where we left off
  "minIterations": 3,
  "maxIterations": 20,
  "completionPromise": "COMPLETE",
  "startedAt": "2026-01-21T18:00:00.000Z",
  "prompt": "Build a REST API...",
  "previousFeedback": {               // Preserved for continuity
    "qualityScore": 7,
    "qualitySummary": "Good progress on core functionality",
    "nextSteps": [
      "Implement caching for performance",
      "Add documentation for public API"
    ],
    "improvements": [
      "Add more edge case tests",
      "Refactor duplicate code in utils module"
    ],
    "blockers": [
      "Need clarification on authentication requirements"
    ],
    "ideas": [
      "Could add CLI flag for verbose output",
      "Consider adding progress indicators"
    ]
  }
}
```

---

### 1. New Command: `resume`

**File:** [src/commands/resume.ts](src/commands/resume.ts)

**Key Features:**
- Validates that a stopped loop exists (state file with `active: false`)
- Extracts all previous feedback and context
- Builds enhanced prompt with resume instructions
- Allows parameter overrides while preserving defaults from state

**Resume Context Builder:**
```typescript
function buildResumeContext(state: LoopState): string {
  // Creates structured prompt with:
  // 1. Notice about resuming at iteration N
  // 2. Instructions to review files, git, tests
  // 3. Breakdown of previous feedback
  // 4. Original task at the end
}
```

### 2. Modified Command: `cancel`

**File:** [src/commands/cancel.ts](src/commands/cancel.ts)

**Before:**
```typescript
// Delete state files
await Promise.all([
  unlink(STATE_FILE).catch(() => {}),
  unlink(SESSION_FILE).catch(() => {}),
]);
```

**After:**
```typescript
// Mark state as inactive (preserves for resume)
state.active = false;
await Bun.write(STATE_FILE, JSON.stringify(state));

// Delete session file (not needed for resume)
await unlink(SESSION_FILE).catch(() => {});

log.message(`Cancelled Ralph loop at iteration ${iteration}`);
log.message("Run 'ralph resume' to continue where you left off.");
```

### 3. Enhanced Loop Runner

**File:** [src/core/loop-runner.ts](src/core/loop-runner.ts)

**Enhanced Cleanup Function:**

```typescript
const cleanup = async (
  currentIteration?: number,
  currentFeedback?: RalphFeedback,
): Promise<void> => {
  // If we have iteration info, preserve the state for resume
  if (currentIteration !== undefined && currentIteration > 0) {
    try {
      const state: LoopState = {
        active: false,              // Mark as inactive
        iteration: currentIteration,
        minIterations: config.minIterations,
        maxIterations: config.maxIterations,
        completionPromise: config.completionPromise,
        startedAt: new Date().toISOString(),
        prompt: config.prompt,
        previousFeedback: currentFeedback,
      };
      await mkdir(dirname(STATE_FILE), { recursive: true });
      await Bun.write(STATE_FILE, stateToJson(state));
    } catch {
      // If we can't write state, just delete it
      await unlink(STATE_FILE).catch(() => {});
    }
  } else {
    // No iteration info, just delete the state file
    // (This happens on successful completion)
    await unlink(STATE_FILE).catch(() => {});
  }
};
```

**SIGINT Handler (Ctrl+C):**
```typescript
process.on("SIGINT", async () => {
  log.warn(pc.yellow(`\nInterrupted at iteration ${iteration}`));
  log.message("Saving state for resume...");
  await cleanup(iteration, previousFeedback);  // Preserves state
  log.message("Run 'ralph resume' to continue where you left off.");
  process.exit(1);
});
```

**Max Iterations Handler:**
```typescript
if (config.maxIterations > 0 && iteration >= config.maxIterations) {
  log.warn(pc.red(`Max iterations (${config.maxIterations}) reached`));
  log.message("Saving state for resume...");
  await cleanup(iteration, previousFeedback);  // Preserves state
  log.message("Run 'ralph resume' to continue where you left off.");
  process.exit(0);
}
```

**Successful Completion Handler:**
```typescript
if (checkCompletionPromise(session, config.completionPromise)) {
  log.success(pc.green(`Completed at iteration ${iteration}!`));
  // Task completed successfully - can delete state
  await cleanup();  // No params = deletes state
  process.exit(0);
}
```

### 4. Schema Updates

**File:** [src/schemas/config.ts](src/schemas/config.ts)

Added two new fields to `LoopConfigSchema`:

```typescript
export const LoopConfigSchema = z.object({
  // ... existing fields ...

  /** Whether this is a resume operation. Default: false */
  isResume: z.boolean().default(false),

  /** Iteration number to resume from. Default: 0 */
  resumeFromIteration: z.number().int().min(0).default(0),
});
```

### 5. CLI Integration

**File:** [src/index.ts](src/index.ts)

```typescript
// Resume command
program
  .command("resume")
  .description("Resume a stopped Ralph loop with context about previous work")
  .option(
    "-n, --min-iterations <number>",
    "Override minimum iterations before checking completion",
  )
  .option(
    "-m, --max-iterations <number>",
    "Override maximum iterations (0 = unlimited)",
  )
  .option(
    "-p, --completion-promise <string>",
    "Override promise phrase to signal completion",
  )
  .option("-a, --agent <name>", "Agent name (default: ralph-wiggum)")
  .action(resumeCommand);
```

---

### Basic Resume

```bash
# Resume with all existing settings from state file
ralph resume
```

**Output:**
```
ℹ  Resuming Ralph loop
   Previous iteration: 12
   Original prompt: Build a REST API for todos with CRUD operations
   Last quality score: 7/10

ℹ  Ralph loop resuming
   Min iterations: 3
   Max iterations: 20
   Completion promise: COMPLETE
   Resuming from iteration: 12
```

### Resume with Parameter Overrides

```bash
# Increase max iterations to continue longer
ralph resume -m 30

# Change the completion promise
ralph resume -p "FINISHED"

# Override minimum iterations
ralph resume -n 5

# Override multiple parameters
ralph resume -n 5 -m 25 -p "DONE"

# Use a different agent
ralph resume -a custom-agent
```

### Complete Workflow Example

```bash
# 1. Start a loop
ralph loop "Build a REST API with tests" -m 20 -p "COMPLETE"

# Work happens...
# Iteration 1, 2, 3, 4, 5...
# Press Ctrl+C at iteration 5

# Output:
# ⚠  Interrupted at iteration 5
#    Saving state for resume...
#    Run 'ralph resume' to continue where you left off.

# 2. Later, resume the work
ralph resume

# Output:
# ℹ  Resuming Ralph loop
#    Previous iteration: 5
#    Original prompt: Build a REST API with tests
#    Last quality score: 6/10
#
# ℹ  Ralph loop resuming
#    Resuming from iteration: 5
#
# ⚙  Iteration 6
# [Agent continues from where it left off...]

# 3. Hit max iterations at 20
# Output:
# ⚠  Max iterations (20) reached
#    Saving state for resume...
#    Run 'ralph resume' to continue where you left off.

# 4. Decide to continue with more iterations
ralph resume -m 30

# Agent continues from iteration 20 to 30

# 5. Eventually completes successfully
# Output:
# ✔  Completed at iteration 24!
# [State file automatically deleted]
```

---

### Resume Workflow

```
User runs: ralph resume -m 30

┌─────────────────────────────────────────────────────────────┐
│ 1. Validate State File Exists                              │
├─────────────────────────────────────────────────────────────┤
│   Read: .kiro/ralph-loop.local.json                         │
│   Check: active === false (stopped loop)                    │
│   Extract:                                                   │
│   ├─ iteration: 12                                          │
│   ├─ prompt: "Build a REST API"                             │
│   ├─ previousFeedback: {...}                                │
│   └─ all other settings                                     │
└─────────────────────────────────────────────────────────────┘
                            ↓
┌─────────────────────────────────────────────────────────────┐
│ 2. Build Enhanced Prompt with Resume Context               │
├─────────────────────────────────────────────────────────────┤
│   RESUME CONTEXT:                                           │
│   ===============                                           │
│   You are resuming at iteration 12                          │
│                                                              │
│   Before continuing:                                         │
│   1. Review files created/modified                          │
│   2. Check git history (git log, git diff)                  │
│   3. Review test results and build artifacts                │
│                                                              │
│   Previous feedback:                                         │
│   - Quality: 7/10 - Good progress                           │
│   - Next steps:                                              │
│     • Implement caching                                      │
│     • Add documentation                                      │
│   - Improvements needed:                                     │
│     • More edge case tests                                   │
│     • Refactor utils module                                  │
│                                                              │
│   Original task: Build a REST API                           │
└─────────────────────────────────────────────────────────────┘
                            ↓
┌─────────────────────────────────────────────────────────────┐
│ 3. Merge Settings (Override + Existing)                    │
├─────────────────────────────────────────────────────────────┤
│   prompt: [Enhanced with resume context]                    │
│   minIterations: 3 (from state)                             │
│   maxIterations: 30 (OVERRIDDEN from -m flag)               │
│   completionPromise: "COMPLETE" (from state)                │
│   isResume: true                                             │
│   resumeFromIteration: 12                                    │
└─────────────────────────────────────────────────────────────┘
                            ↓
┌─────────────────────────────────────────────────────────────┐
│ 4. Start Loop Runner in Resume Mode                        │
├─────────────────────────────────────────────────────────────┤
│   ✓ Skip cleanup (preserve state file)                      │
│   ✓ Set iteration = 12 (not 0)                              │
│   ✓ Display "Ralph loop resuming"                           │
│   ✓ Show "Resuming from iteration: 12"                      │
└─────────────────────────────────────────────────────────────┘
                            ↓
┌─────────────────────────────────────────────────────────────┐
│ 5. Execute Iterations                                       │
├─────────────────────────────────────────────────────────────┤
│   Iteration 13:                                              │
│   ├─ Write state (iteration: 13, active: true)              │
│   ├─ Run kiro-cli with enhanced prompt                      │
│   ├─ Extract feedback from session                          │
│   └─ Check completion (no) / max (no)                       │
│                                                              │
│   Iteration 14-29: [continue...]                            │
│                                                              │
│   Iteration 30:                                              │
│   ├─ Write state (iteration: 30, active: true)              │
│   ├─ Run kiro-cli                                           │
│   ├─ Extract feedback                                       │
│   └─ Check max iterations (30 >= 30) → YES                  │
│       ├─ Mark state as inactive                             │
│       ├─ Save with iteration 30 and feedback                │
│       └─ Exit with resume message                           │
└─────────────────────────────────────────────────────────────┘
```

### State Lifecycle Diagram

```
┌──────────────┐
│ New Loop     │
│ ralph loop   │
└──────┬───────┘
       │
       ├─► Delete any existing state (cleanup)
       │
       ▼
┌──────────────────────────────────────────┐
│ Loop Running                             │
│ State: { active: true, iteration: N }    │
└──────┬───────────────────────────────────┘
       │
       ├──── Ctrl+C ────────────────────┐
       │                                 │
       ├──── Max Iterations ─────────────┤
       │                                 │
       ├──── ralph cancel ───────────────┤
       │                                 │
       │                                 ▼
       │              ┌────────────────────────────────────┐
       │              │ State Preserved                    │
       │              │ { active: false, iteration: N,     │
       │              │   previousFeedback: {...} }        │
       │              └──────┬─────────────────────────────┘
       │                     │
       │                     │ ralph resume
       │                     │
       │                     ▼
       │              ┌────────────────────────────────────┐
       │              │ Resume Loop                        │
       │              │ Starts at iteration N              │
       │              │ Enhanced prompt with context       │
       │              └──────┬─────────────────────────────┘
       │                     │
       │                     └─► Back to "Loop Running"
       │
       ├──── Task Complete (promise detected) ───┐
       │                                          │
       ▼                                          ▼
┌────────────────────┐              ┌────────────────────┐
│ State Deleted      │              │ State Deleted      │
│ (cleanup with no   │              │ (successful        │
│ iteration info)    │              │ completion)        │
└────────────────────┘              └────────────────────┘
```

---

### New Files

- **`src/commands/resume.ts`** - Complete resume command implementation
- **`RESUME_IMPLEMENTATION.md`** - This comprehensive documentation

### Modified Files

- **`src/commands/cancel.ts`**
  - Changed to preserve state instead of deleting
  - Marks `active: false` for resume capability
  - Displays resume instructions

- **`src/core/loop-runner.ts`**
  - Enhanced `cleanup()` function with iteration/feedback params
  - Preserves state on Ctrl+C with current iteration
  - Preserves state when max iterations reached
  - Only deletes state on successful completion
  - Supports `isResume` and `resumeFromIteration` config

- **`src/schemas/config.ts`**
  - Added `isResume: boolean` field
  - Added `resumeFromIteration: number` field

- **`src/index.ts`**
  - Import `resumeCommand`
  - Register `resume` command with CLI options
  - Updated file header comment

- **`src/commands/index.ts`**
  - Export `resumeCommand`

- **`README.md`**
  - Added complete `ralph resume` documentation
  - Updated `ralph cancel` documentation
  - Added state preservation notes
  - Updated Quick Start with resume example

### Removed Files

- **`RESUME_FEATURE.md`** - Merged into this document
- **`RESUME_FIX.md`** - Merged into this document

---

## Use Cases

1. **Interrupted Loops (Ctrl+C)**
   - Developer needs to stop work temporarily
   - System needs restart
   - Want to review current progress before continuing

2. **Hit Iteration Limit**
   - Initially set conservative max iterations
   - Task needs more iterations than estimated
   - Want to continue with extended limit

3. **Cancelled Loops**
   - Realized different parameters needed
   - Want to restart with different completion criteria
   - Testing/debugging the loop itself

4. **Parameter Adjustment**
   - Change iteration limits mid-task
   - Modify completion promise
   - Switch to different agent

5. **Fresh Re-evaluation**
   - Let agent reassess work with clean perspective
   - Provide updated context about blockers
   - Continue after external changes (manual code edits, etc.)

---

## Future Enhancements

Potential improvements for consideration:

1. **Reset Iterations Flag**
   ```bash
   ralph resume --reset-iterations
   # Start count from 1 but keep feedback
   ```

2. **Diff Display**
   ```bash
   ralph resume --show-diff
   # Display git diff since last iteration
   ```

3. **Interactive Mode**
   ```bash
   ralph resume --interactive
   # Review/edit resume context before starting
   ```

4. **Custom Instructions**
   ```bash
   ralph resume --instruction "Focus on error handling"
   # Append custom guidance to resume context
   ```

5. **Resume History**
   - Track how many times a loop has been resumed
   - Store timestamps of each resume
   - Display resume history with `ralph status`

6. **Git Log Integration**
   - Automatically include relevant git log in resume context
   - Show commits since loop started

---

## Testing

The implementation includes comprehensive validation:

```bash
# All checks should pass:
# ✓ resumeCommand imported and registered in CLI
# ✓ resumeCommand exported from commands module
# ✓ resume.ts file exists with complete implementation
# ✓ isResume field added to LoopConfigSchema
# ✓ resumeFromIteration field added to LoopConfigSchema
# ✓ loop-runner checks isResume flag
# ✓ loop-runner uses resumeFromIteration
# ✓ Cancel preserves state (marks inactive)
# ✓ Ctrl+C preserves state with iteration and feedback
# ✓ Max iterations preserves state
# ✓ Successful completion deletes state
```

---

## Summary

✅ **Resume Command**: Fully functional with enhanced context
✅ **State Preservation**: Automatic on interruption/cancellation
✅ **Parameter Overrides**: All loop settings can be modified on resume
✅ **Iteration Continuity**: Picks up exactly where it left off
✅ **Feedback Bridge**: Previous iteration feedback preserved and provided
✅ **Clean Completion**: State automatically deleted on success
✅ **User Experience**: Clear messages guide users to resume

The `ralph resume` feature transforms Ralph from a one-shot tool into a persistent, resumable development assistant that can handle long-running tasks across multiple sessions.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -50,4 +55,23 @@ program
...
+	.option(
+		"-n, --min-iterations <number>",
+		"Override minimum iterations before checking completion",
+	)
+	.option(
+		"-m, --max-iterations <number>",
+		"Override maximum iterations (0 = unlimited)",
+	)
+	.option(
+		"-p, --completion-promise <string>",
```
Comment (theagenticguy) [inline src/index.ts:71]: should resume have a different completion promise string?

Comment (harmjeff) [reply src/index.ts:71]: It is optional if you want to change the goal

**Review by theagenticguy — APPROVED** [review]
Fixed Biome lint errors (noImplicitAnyLet) and flaky agent override test timeout. All 7 CI checks passing.
